control/pid.go

```diff
@@ -50,41 +53,102 @@ func (p *basicPID) Next(ctx context.Context, x []*Signal, dt time.Duration) ([]*
 	p.mu.Lock()
 	defer p.mu.Unlock()
 	if p.tuning {
-		out, done := p.tuner.pidTunerStep(math.Abs(x[0].GetSignalValueAt(0)), p.logger)
-		if done {
-			p.kD = p.tuner.kD
-			p.kI = p.tuner.kI
-			p.kP = p.tuner.kP
-			p.logger.Info("\n\n-------- ***** PID GAINS CALCULATED **** --------")
-			p.logger.CInfof(ctx, "Calculated gains are p: %1.6f, i: %1.6f, d: %1.6f", p.kP, p.kI, p.kD)
-			p.logger.CInfof(ctx, "You must MANUALLY ADD p, i and d gains to the robot config to use the values after tuning\n\n")
-			p.tuning = false
+		// Multi Input/Output Implementation
+		if p.useMulti {
+
+			// For each PID Set and its respective Tuner Object, Step through an iteration of tuning until done.
+			for i := 0; i < len(p.PIDSets); i++ {
+
+				out, done := p.tuners[i].pidTunerStep(math.Abs(x[0].GetSignalValueAt(i)), p.logger)
+				if done {
+
+					p.PIDSets[i].D = p.tuners[i].kD
+					p.PIDSets[i].I = p.tuners[i].kI
+					p.PIDSets[i].P = p.tuners[i].kP
+					p.logger.Info("\n\n-------- ***** PID GAINS CALCULATED **** --------")
+					p.logger.CInfof(ctx, "Calculated gains are p: %1.6f, i: %1.6f, d: %1.6f", p.PIDSets[i].P, p.PIDSets[i].I, p.PIDSets[i].D)
+					p.logger.CInfof(ctx, "You must MANUALLY ADD p, i and d gains to the robot config to use the values after tuning\n\n")
+					p.tuning = false
+				}
+				p.y[0].SetSignalValueAt(i, out)
+			}
+		} else {
+			// Single Input/Output Implementation
+			out, done := p.tuner.pidTunerStep(math.Abs(x[0].GetSignalValueAt(0)), p.logger)
+			if done {
+				p.kD = p.tuner.kD
+				p.kI = p.tuner.kI
+				p.kP = p.tuner.kP
+				p.logger.Info("\n\n-------- ***** PID GAINS CALCULATED **** --------")
+				p.logger.CInfof(ctx, "Calculated gains are p: %1.6f, i: %1.6f, d: %1.6f", p.kP, p.kI, p.kD)
+				p.logger.CInfof(ctx, "You must MANUALLY ADD p, i and d gains to the robot config to use the values after tuning\n\n")
+				p.tuning = false
+			}
+			p.y[0].SetSignalValueAt(0, out)
 		}
-		p.y[0].SetSignalValueAt(0, out)
+
 	} else {
-		dtS := dt.Seconds()
-		pvError := x[0].GetSignalValueAt(0)
-		p.int += p.kI * pvError * dtS
-		switch {
-		case p.int >= p.satLimUp:
-			p.int = p.satLimUp
-		case p.int <= p.satLimLo:
-			p.int = p.satLimLo
-		default:
-		}
-		deriv := (pvError - p.error) / dtS
-		output := p.kP*pvError + p.int + p.kD*deriv
-		p.error = pvError
-		if output > p.limUp {
-			output = p.limUp
-		} else if output < p.limLo {
-			output = p.limLo
+
+		// Multi Input/Output Implementation
+		if p.useMulti {
+
+			for i := 0; i < len(p.PIDSets); i++ {
+				output := calculateSignalValue(p, x, dt, i)
```

Comment:
can this function be used for the single input case?

Reply:
Not going to expand this for the single input case. will hold off on making large changes to the single input case

Reply:
eventually I think yes and just index 0 any time it's single input. but that does seem like a problem for a future PR